### ISSUE TITLE
Is it possible to view the core web vitals for each page?

### ISSUE THREAD

**Issue by emersonthis** [TGiles/auto-lighthouse#128]
This is a really cool tool!

I'm interested to see the metrics for each individual page that was crawled. The generated report appears to be an aggregate of all the pages. Is that correct? I'd love to see a CSV file with each page/link as a row. Is something like this already possible.

**Comment by TGiles** [comment]
@emersonthis 
Thanks for the kind words, I appreciate seeing people get use out of this tool! Let me see if I understand your question correctly...it sounds like you want to see each page as a separate report. If I'm correct in this assumption, generating a report for each page is the tool's default behavior when given a domain (or domains). If you wish to have these reports in CSV form, run the tool on your requested domains and pass the `--format csv` to have each report generated as a CSV. There is no current way to aggregate reports into a single report, especially since each user would want a different way to aggregate reports (at least that's what I see in my head).

To sum it up, run the tool in CSV mode by passing `--format csv` when you run the tool. This will generate a CSV that has the individual page, the category of metrics, the individual metric, and the score for the metric.

**Comment by emersonthis** [comment]
Huh! Your assumption is correct. This is great to hear that the tool _already_ does what I want. But somehow I'm still not understanding how to do it. I think I was confused because I was passing a URL via the `-u` parameter, and that returns a single report which I thought (incorrectly) was an aggregation of all the pages. 

```sh
$ npx auto-lighthouse -u https://example.com --format csv
```

But I'm still confused about how to pass a domain! Here's what I've tried:
```sh
$ npx auto-lighthouse -u example.com --format csv
--> Invalid URL
```

```sh
$ npx auto-lighthouse example.com --format csv
--> No URL given, quitting!
```

I'm sure I'm overlooking something obvious, but can give me an example of how to correctly pass a domain?

**Comment by TGiles** [comment]
Ah okay, I see your issue. I use simplecrawler under the hood to crawl domains, and simplecrawler needs a URL that has a protocol at the beginning of the URL. So if the domain you wanted to test is `www.myDomain.com`, you would run the following command:

```bash
$ npx auto-lighthouse -u https://www.myDomain.com --format csv
```

If the domain you are testing is only one page, which I'm guessing `https://example.com` is, the tool will only generate one report. There should also be output in the terminal showing which pages have been found, which pages are being added to queue, and which pages are currently being used to generate reports.

I'll work on improving the URL flag part of the experience, I'll add a check to see if a protocol is in the URL strings and if not, add a `https://`. I think simplecrawler can handle the parsing from there with no issue (say if the site you were testing was `http` only, I think simplecrawler can switch the protocol to `http` on run).

Please let me know if that answers your question!

**Comment by emersonthis** [comment]
@TGiles I've been running the command just as you instructed, and still getting a single report page for the homepage (only)... I think I might have found the issue. The real site I'm trying to test has a robots.txt file with:

```
User-agent: *
Dissallow: /
```

Would this explain why the crawler is only finding the homepage? I tested on a different live site and it seems to be working as expected. If the issue is the robots.txt file, is there any way we can override that? We're trying to do performance analysis on a site that's in a QA environment.

Assuming I can get multiple pages to work, is there any way to get the output in a single CSV file instead of multiple files? We're trying to look across hundreds of pages and identify the ones that are least performant. This would be a lot easier if I can sort rows in a spreadsheet instead of opening up each file one-by-one.

**Comment by TGiles** [comment]
Ah a robots.txt issue. Yeah by default simplecrawler will respect robots.txt rules and `nofollow` rules in meta tags. I can easily expose  the option to ignore robots txt to the CLI, solving one part of your issue. However, me solving the robots issue is minuscule in comparison to solving the single file issue.

Lighthouse, by default, has no aggregation feature. I'm sure there are paid services that have figured this out, but I have not implemented yet. I'll work on a proof of concept to see if I can get an aggregation that I'm comfortable with, at least in CSV mode. I think it would be impossible in the other output mode (not that you're asking for that, just thinking out loud).

**Comment by emersonthis** [comment]
@TGiles Want to "divide and conquer"? If you can expose the option to ignore robots.txt I can take a pass at bundling the reports into a single CSV file. I'll probably start by writing it as a standalone script that reads the individual report files generated by the existing tool, which should be pretty easy. Then if you like the output from that, you could consider lumping in into the main project with an additional flag or something. How does this sound?

**Comment by TGiles** [comment]
@emersonthis yeah, that sounds like a fine plan to me. I'll set up a new branch (probably call it `feat/add_aggregation_for_csv`) and push the robots.txt flag to that branch. I also have a proof of concept for appending the files together, just need a way to remove the headers of the CSVs after I add the first file. If you end up figuring it out before me though, fine with me.

**Comment by TGiles** [comment]
@emersonthis 
Since you opened this issue, I also discovered a typo in my threads option logic! I've gone ahead and exposed an option for respecting the robots.txt rules, `--respectRobots`. Additionally, I've fixed your previous pain point of using the URL option with no protocol. When you have your changes available, please create a pull request into the `feat/add_aggregation_for_csv` branch. Thanks!

**Comment by emersonthis** [comment]
Cool! I haven't looked at the branch yet, but I want to make sure I understand the expected behavior. The `--respectRobots` flat is a boolean? If true, it respects robots.txt. If false, it ignores and crawls it anyway?

**Comment by TGiles** [comment]
The `--respectRobots` works exactly as you have described. I also have my initial attempt of aggregating the reports together, along with a simple test to ensure it works as expected. Currently working on the issue of, I need to create two aggregate reports, one for mobile and one for desktop reports. Otherwise, it is coming along well.

**Comment by emersonthis** [comment]
Cool! Sounds like you’re moving fast! If/when you’re ready, let me know the options to test the new aggregated csv and I’ll give it a try. This was already a cool tool, but this feature feels like a game-changer!

**Comment by TGiles** [comment]
@emersonthis the `feat/add_aggregation_for_csv` branch should be ready for end user testing. I've written some best case and simple error tests for the aggregate logic, but feel free to review that as well. The default behavior when using the tool in CSV mode is to create two aggregate reports, one for mobile and one for desktop. You should now also be able to use the `-u` flag with any type of URL (with or without http/https). I don't have a test case for the `--respectRobots` flag, so I'll be leaning on your testing to see if it performs as expected, otherwise I'll have to file something with simplecrawler (I suspect the `--respectRobots` will work as expected, as long as I've assigned the value correctly to simplecrawler).

**Comment by TGiles** [comment]
welp just kidding, apparently there a testing error that appears on the build server, but not my machine. You should still be able to use it, but I'll need to figure out why `fs` is throwing errors, instead of just logging them in this particular case.

**Comment by emersonthis** [comment]
@TGiles I just checked out the branch and tested with `--respectRobots false` and it appears to be working perfectly. Thanks!

As for the csv aggregation, it looks like in this branch the aggregation becomes the default behavior when the the format is CSV? Is that correct? No additional flags, etc?

**Comment by TGiles** [comment]
@emersonthis glad to hear that I've fed simplecrawler the correct configuration. As for your second point, you are entirely correct. My gut says that creating the aggregate reports should happen automatically in CSV mode, since I see the main use case of CSV being what you originally requested; being able to compare scores across your entire domain in a way that is much easier than opening reports one by one. I've exported the functionality of the aggregate logic though, so that it can be used outside of the interactive mode. Figured it would be helpful.

**Comment by emersonthis** [comment]
>  I've exported the functionality of the aggregate logic though, so that it can be used outside of the interactive mode. Figured it would be helpful.

Can you elaborate on that? I just did a test run and it seems to still be writing individual CSV files for each page. Is there there a different workflow I should follow to get the aggregation working?

**Comment by TGiles** [comment]
Yeah sure. If you wanted to create new aggregate reports, you could use a snippet like this:

```js

const runner = require('auto-lighthouse');
const aggregateFunction = runner.aggregateCSVReports;

aggregateFunction('someFilePath');
```
If I've set it up correctly, this will generate new aggregate CSVs based on some directory (as long as the directory was created by running the auto-lighthouse tool).

To elaborate on your question: the tool will still write individual CSV files (since it uses these CSVs to create the aggregate CSV). As long as you're using the new feature branch to test, and using the `--format csv` mode, the tool should output both individual CSVs and two aggregate CSVs that have the timestamp plus ".desktopAggregate" or ".mobileAggregate" or something like that as the file name. 

Please let me know if it's not working the way you expect, I want to hammer out the experience before I create a new release!

**Comment by emersonthis** [comment]
I'm still waiting for the individual CSVs to finish getting written. It's taking a long time because the site I'm scanning is pretty large. So I haven't gotten to the aggregated CSV file yet, but I already have a thought about the experience: it would be great if there was eventually a way to generate the aggregated file without waiting for each individual file to write.

I know this is probably more annoying to implement because it would require a bit of refactoring. But for a large site it would save a lot of time. The site I'm testing returns thousands of URLs, and so it's been running for over an hour and at least half of that time is writing the individual CSVs. I'm thinking that if there was a way to store that data in memory it could half the total time. I'll follow-up again whenever the script finishes running (I should have started testing on a smaller site 🤪).

**Comment by TGiles** [comment]
Just throwing my thoughts out here, but I think the experience you want is impossible. In order to have a CSV with all the reports, all the reports need to finish. Since Lighthouse is running an audit on every single page, twice, this takes time, especially if there a large amount of content on the page. As far as I know, Lighthouse doesn't have a stream-like capability for running reports. In the current implementation, I'm able to generate reports asynchronously but I don't know when the reports are done until they are done. I could take my current approach of generating the aggregate reports, but apply it so that it adds to the write stream when a report is done, but the aggregate file would not be viewable until the write stream is closed (if I understand Node's file system streams correctly).

In summary, my gut says that the experience you're asking for is impossible. Glad to be proven wrong, but I don't see a way around.

**Comment by emersonthis** [comment]
That's interesting. I assumed lighthouse was returning data... that was then getting written to files. I didn't realize it was writing the files too. I'll do a little research to understand that API better.

**Comment by emersonthis** [comment]
My test run finally completed. In the future I will definitely experiment on smaller sites! The aggregated reports seem to have been generated correctly. However the command ended with the following errors:

```
Error: not opened
    at WebSocket.send (/Users/emerson/Code/auto-lighthouse/node_modules/ws/lib/WebSocket.js:344:18)
    at CriConnection.sendRawMessage (/Users/emerson/Code/auto-lighthouse/node_modules/lighthouse/lighthouse-core/gather/connections/cri.js:167:14)
    at CriConnection.sendCommand (/Users/emerson/Code/auto-lighthouse/node_modules/lighthouse/lighthouse-core/gather/connections/connection.js:66:10)
    at Driver._innerSendCommand (/Users/emerson/Code/auto-lighthouse/node_modules/lighthouse/lighthouse-core/gather/driver.js:397:29)
    at /Users/emerson/Code/auto-lighthouse/node_modules/lighthouse/lighthouse-core/gather/driver.js:359:35
    at new Promise (<anonymous>)
    at Driver.sendCommandToSession (/Users/emerson/Code/auto-lighthouse/node_modules/lighthouse/lighthouse-core/gather/driver.js:350:12)
    at Driver.sendCommand (/Users/emerson/Code/auto-lighthouse/node_modules/lighthouse/lighthouse-core/gather/driver.js:377:17)
    at Driver._getOrCreateIsolatedContextId (/Users/emerson/Code/auto-lighthouse/node_modules/lighthouse/lighthouse-core/gather/driver.js:1059:45)
    at Driver.evaluateAsync (/Users/emerson/Code/auto-lighthouse/node_modules/lighthouse/lighthouse-core/gather/driver.js:431:57) {
  friendlyMessage: undefined
}
TypeError: Cannot read property 'categories' of undefined
    at Function.generateReportCSV (/Users/emerson/Code/auto-lighthouse/node_modules/lighthouse/lighthouse-core/report/report-generator.js:72:37)
    at /Users/emerson/Code/auto-lighthouse/node_modules/lighthouse/lighthouse-core/report/report-generator.js:105:32
    at Array.map (<anonymous>)
    at Function.generateReport (/Users/emerson/Code/auto-lighthouse/node_modules/lighthouse/lighthouse-core/report/report-generator.js:98:32)
    at processResults (/Users/emerson/Code/auto-lighthouse/lighthouse_runner.js:96:33)
    at /Users/emerson/Code/auto-lighthouse/lighthouse_runner.js:62:21
    at async processReports (/Users/emerson/Code/auto-lighthouse/lighthouse_runner.js:54:13)
    at async /Users/emerson/Code/auto-lighthouse/lighthouse_runner.js:131:9
    at async Promise.all (index 0)
    at async /Users/emerson/Code/auto-lighthouse/lighthouse_runner.js:213:13
TypeError: Cannot read property 'categories' of undefined
    at Function.generateReportCSV (/Users/emerson/Code/auto-lighthouse/node_modules/lighthouse/lighthouse-core/report/report-generator.js:72:37)
    at /Users/emerson/Code/auto-lighthouse/node_modules/lighthouse/lighthouse-core/report/report-generator.js:105:32
    at Array.map (<anonymous>)
    at Function.generateReport (/Users/emerson/Code/auto-lighthouse/node_modules/lighthouse/lighthouse-core/report/report-generator.js:98:32)
    at processResults (/Users/emerson/Code/auto-lighthouse/lighthouse_runner.js:96:33)
    at /Users/emerson/Code/auto-lighthouse/lighthouse_runner.js:62:21
    at async processReports (/Users/emerson/Code/auto-lighthouse/lighthouse_runner.js:54:13)
    at async /Users/emerson/Code/auto-lighthouse/lighthouse_runner.js:131:9
    at async Promise.all (index 0)
    at async /Users/emerson/Code/auto-lighthouse/lighthouse_runner.js:213:13
Done with reports!
```

Also, in my mind I was imaging the metrics transposed into columns so that each row is a single page. I can think of a few use-cases that would make that more useful to work with. But there also might be other issues I'm not thinking about that make this "stacked" format better. I'll give it some more thought.

**Comment by TGiles** [comment]
Seems like there was some internal error with Lighthouse report generation. Unfortunately, doesn't appear to be anything on my end. The crawler might have grabbed a URL that Lighthouse could not use, based on the first part of your error message "Error: not opened." In any case, I'm going to go ahead and merge my PR adding the CSV feature which will close this issue.

Please open new issues if you have other feature requests/bug fixes that I need to look at!

**Comment by TGiles** [comment]
Closed by #130
